src/ucs/sys/topo/base/topo.c

```diff
@@ -790,3 +793,122 @@ const char *ucs_topo_resolve_sysfs_path(const char *dev_path, char *path_buffer)
+{
+    static const char *low_level_dev_prefix = "lower_";
+    ucs_string_buffer_t *dev_path           = ctx;
+    char sysfs_path[PATH_MAX];
+
+    if (!strncmp(entry->d_name, low_level_dev_prefix,
+                 strlen(low_level_dev_prefix))) {
+        ucs_string_buffer_appendf(dev_path, "/%s", entry->d_name);
+        if (ucs_topo_resolve_sysfs_path(ucs_string_buffer_cstr(dev_path),
+                                        sysfs_path)) {
+            return UCS_ERR_CANCELED;
+        }
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_topo_iface_parse_bond_dev(const struct dirent *entry, void *ctx)
+{
+    ucs_string_buffer_t bonding_path = UCS_STRING_BUFFER_INITIALIZER;
+    static const char *low_level_dev_prefix = "lower_";
+    ucs_string_buffer_t *dev_path           = ctx;
+
+    if (!strncmp(entry->d_name, low_level_dev_prefix,
+                 strlen(low_level_dev_prefix))) {
+        ucs_string_buffer_appendf(dev_path, "/%s", entry->d_name);
+        ucs_string_buffer_appendf(&bonding_path, "%s/%s",
+                                  ucs_string_buffer_cstr(dev_path),
+                                  UCS_BONDING_AD_NUM_PORTS);
+        if (!access(ucs_string_buffer_cstr(&bonding_path), F_OK)) {
+            ucs_string_buffer_cleanup(&bonding_path);
+            return UCS_ERR_CANCELED;
+        }
+        ucs_string_buffer_cleanup(&bonding_path);
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucs_topo_iface_deep_readdir(const char *path, ucs_sys_readdir_cb_t cb,
+                            ucs_string_buffer_t* dev_path,
+                            int max_depth) {
+    int depth;
+    ucs_status_t status = UCS_OK;
+    for (depth = 0; depth < max_depth; ++ depth) {
+        status = ucs_sys_readdir(ucs_string_buffer_cstr(dev_path),
+                                 cb, dev_path);
+        if (status == UCS_ERR_CANCELED) {
+            break;
+        }
+    }
+
+    return status;
+}
+
+const char *ucs_topo_iface_get_sysfs_path(const char *dev_name,
+                                          const char *dev_type_root,
+                                          char *path_buffer)
+{
+    ucs_string_buffer_t dev_path = UCS_STRING_BUFFER_INITIALIZER;
+    const char *sysfs_path;
+    ucs_status_t status;
+
+    /* Deep search and return a correct device sysfs path:
+     * 1) For regular device, use regular sysfs form.
+     * 2) For virtual device (RoCE LAG/VLAN), search for symbolic link of the
+     *    form "lower_*" */
+    ucs_string_buffer_appendf(&dev_path, "%s/%s", dev_type_root, dev_name);
+
+    status = ucs_topo_iface_deep_readdir(ucs_string_buffer_cstr(&dev_path),
+                                         uct_topo_iface_parse_virtual_dev,
+                                         &dev_path, UCS_TOPO_MAX_READ_DIR_DEPTH);
+    if (status != UCS_ERR_CANCELED) {
+        ucs_string_buffer_cleanup(&dev_path);
+        return NULL;
+    }
+
```

Comment:
move this code `ucs_topo_iface_deep_readdir`

Reply:
Hi @shasson5, I moved the code and also moved back the ucs_topo_iface_get_sysfs_path to tcp_iface.c and restore to the original name.

Reply:
@shasson5. Please review the new commits. Thanks